mmf/models/unit/matcher.py

```diff
@@ -0,0 +1,166 @@
+        out_prob = self.norm(
+            outputs["pred_logits"].flatten(0, 1)
+        )  # [batch_size * num_queries, num_classes]
+        out_bbox = outputs["pred_boxes"].flatten(0, 1)  # [batch_size * num_queries, 4]
+
+        # Also concat the target labels and boxes
+        tgt_ids = torch.cat([v["labels"] for v in targets])
+        tgt_bbox = torch.cat([v["boxes"] for v in targets])
+
+        # Compute the classification cost. Contrary to the loss, we don't use the NLL,
+        # but approximate it in 1 - proba[target class].
+        # The 1 is a constant that doesn't change the matching, it can be ommitted.
+        cost_class = -out_prob[:, tgt_ids]
+
+        # Compute the L1 cost between boxes
+        cost_bbox = torch.cdist(out_bbox, tgt_bbox, p=1)
+
+        # Compute the giou cost betwen boxes
+        cost_giou = -generalized_box_iou(
+            box_cxcywh_to_xyxy(out_bbox).float(), box_cxcywh_to_xyxy(tgt_bbox)
+        )
+
+        # Final cost matrix
+        C = (
+            self.cost_bbox * cost_bbox
+            + self.cost_class * cost_class
+            + self.cost_giou * cost_giou
+        )
+        C = C.view(bs, num_queries, -1).cpu()
+
+        sizes = [len(v["boxes"]) for v in targets]
+        indices = [
+            linear_sum_assignment(c[i]) for i, c in enumerate(C.split(sizes, -1))
+        ]
+        return [
+            (
+                torch.as_tensor(i, dtype=torch.int64),
+                torch.as_tensor(j, dtype=torch.int64),
+            )
+            for i, j in indices
+        ]
+
+
+class SequentialMatcher(nn.Module):
+    def forward(self, outputs, targets):
+        return [(torch.arange(len(tgt["labels"])),) * 2 for tgt in targets]
+
+
+class LexicographicalMatcher(nn.Module):
+    def __init__(self, lexic="acb"):
+        super().__init__()
+        self.lexic = lexic
+
+    def forward(self, outputs, targets):
+        indices = []
+        for tgt in targets:
+            tgt_cls, tgt_box = tgt["labels"], tgt["boxes"]
+            area = tgt_box[:, 2] * tgt_box[:, 3]
+            if self.lexic == "acb":
+                search_list = [
+                    (-a, cl, b)
+                    for cl, a, b in zip(
+                        tgt_cls.tolist(), area.tolist(), tgt_box.tolist()
+                    )
+                ]
+            else:
+                search_list = [
+                    (cl, -a, b)
+                    for cl, a, b in zip(
+                        tgt_cls.tolist(), area.tolist(), tgt_box.tolist()
+                    )
+                ]
+            # argsort from https://stackoverflow.com/questions/3382352/equivalent-of-numpy-argsort-in-basic-python
+            j = sorted(range(len(search_list)), key=search_list.__getitem__)
+            j = torch.as_tensor(j, dtype=torch.int64)
+            i = torch.arange(len(j), dtype=j.dtype)
+            indices.append((i, j))
+        return indices
+
+
```

Comment:
Do we need others if we are only going to use HungarianMatcher always?

Reply:
(updated to keep only HungarianMatcher)